File: IPWorks MQ Samples/AzureRelay Sender/c++/azurerelaysender.cpp
/*
 * IPWorks MQ 2024 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks MQ in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworksmq
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */


#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "../../include/ipworksmq.h"


#define LINE_LEN 256

char input[LINE_LEN + 1];
bool dataInReceived;

const char* prompt(const char* prompt) {
	printf("%s: ", prompt);
	fgets(input, LINE_LEN, stdin);
	input[strlen(input) - 1] = '\0';
	if (strlen(input) == 0) {
		strncpy(input, "", LINE_LEN);
		input[LINE_LEN] = '\0';
	}
	return input;
}
const char* prompt(const char* prompt, const char* defaultVal) {
	printf("%s [%s]: ", prompt, defaultVal);
	fgets(input, LINE_LEN, stdin);
	input[strlen(input) - 1] = '\0';
	if (strlen(input) == 0) {
		strncpy(input, defaultVal, LINE_LEN);
		input[LINE_LEN] = '\0';
	}
	return input;
}

class MyAzurerelaysender : public AzureRelaySender
{
	virtual int FireSSLServerAuthentication(AzureRelaySenderSSLServerAuthenticationEventParams *e)
	{
		if (!e->Accept)
		{
			printf("Server provided the following certificate: \n");
			printf("Subject: %s \n", e->CertSubject);
			printf("Issuer: %s \n", e->CertIssuer);
			printf("The following problems have been determined for this certificate:  %s \n", e->Status);
			printf("Would you like to continue or cancel the connection?  [y/n] ");

			char response[LINE_LEN];

			fgets(response, LINE_LEN, stdin);
			response[strlen(response) - 1] = '\0';
			if (strcmp(response, "y") == 0)
				e->Accept = true;
		}

		return 0;
	}

	virtual int FireDataIn(AzureRelaySenderDataInEventParams* e) {
		printf("Received: %s\r\n",e->Text);
		dataInReceived = true;
		return 0;
	}

	virtual int FireConnectionConnected(AzureRelayReceiverConnectionConnectedEventParams *e) {
		printf("[%i] connected\r\n", e->ConnectionId);
		return 0;
	}

	virtual int FireConnectionStatus(AzureRelaySenderConnectionStatusEventParams *e) {
		printf("%s\r\n",e->ConnectionEvent);
		return 0;
	}

	virtual int FireDisconnnected(AzureRelayReceiverDisconnectedEventParams* e) {
		printf("Disconnected\n");
		return 0;
	}
};

int main()
{
	MyAzurerelaysender azurerelaysender1;

	printf("*************************************************************************\n");
	printf("* This demo shows how to use the AzureRelaySender component to connect  *\n");
	printf("* and send messages to an AzureRelayReceiver.                           *\n");
	printf("*************************************************************************\n\n");

	// Configure Azure Relay credentials
	azurerelaysender1.SetAccessKey(prompt("Set Access Key"));
	azurerelaysender1.SetAccessKeyName(prompt("Set Access Key Name", "RootManageSharedAccessKey"));
	azurerelaysender1.SetNamespaceAddress(prompt("Set Namespace Address"));
	azurerelaysender1.SetHybridConnection("hc1");
	printf("\n");

	int ret_code = azurerelaysender1.Connect();

	if (ret_code)
	{
		printf("Error connecting: %i - %s\n", ret_code, azurerelaysender1.GetLastError());
		goto done;
	}

	char command[LINE_LEN];
	while (true)
	{
		dataInReceived = false;
		printf("\nPlease input command: \r\n- 1 Send Data \r\n- 2 Exit \r\n");
		printf(">");

		fgets(command, LINE_LEN, stdin);
		command[strlen(command) - 1] = '\0';

		if (!strcmp(command, "1"))
		{
			char text[LINE_LEN];
			printf("Please enter data to send: ");
			fgets(text, LINE_LEN, stdin);
			text[strlen(text) - 1] = '\0';
			ret_code = azurerelaysender1.SendText(text);
			if (ret_code)
			{
				printf("Sending failed: %i - %s\n", ret_code, azurerelaysender1.GetLastError());
			}
			else
			{
				printf("Waiting for response...\n");
				while (!dataInReceived)
					azurerelaysender1.DoEvents();
			}
		}
		else if (!strcmp(command, "2"))
		{
			goto done;
		}
		else
		{
			printf("Command not recognized.\n");
		}
	}

done:
	if (azurerelaysender1.GetConnected())
	{
		azurerelaysender1.Disconnect();
	}
	printf("Exiting... (press enter)\n");
	getchar();

	return 0;
}

File: IPWorks MQ Samples/AzureRelay Sender/python/azurerelaysender.py
# 
# IPWorks MQ 2024 Python Edition - Sample Project
# 
# This sample project demonstrates the usage of IPWorks MQ in a 
# simple, straightforward way. It is not intended to be a complete 
# application. Error handling and other checks are simplified for clarity.
# 
# www.nsoftware.com/ipworksmq
# 
# This code is subject to the terms and conditions specified in the 
# corresponding product license agreement which outlines the authorized 
# usage and restrictions.
# 

import sys
import string
from ipworksmq import *

input = sys.hexversion < 0x03000000 and raw_input or input


def ensureArg(args, prompt, index):
    if len(args) <= index:
        while len(args) <= index:
            args.append(None)
        args[index] = input(prompt)
    elif args[index] is None:
        args[index] = input(prompt)



dataReceived = False

def fireError(e):
  print("ERROR: " %e.message)

def fireSSLServerAuthentication(e):
  if (not e.accept):
    print("Server provided the following certificate: \n")
    print("Subject: %s \n", e.cert_subject)
    print("Issuer: %s \n", e.cert_issuer)
    print("The following problems have been determined for this certificate:  %s \n", e.status)
    response = input("Would you like to continue or cancel the connection?  [y/n] ")
    if (response =="y"):
      e.accept = True

def fireConnectionStatus(e):
  print(e.connection_event)

def fireDataIn(e):
  global dataReceived
  dataReceived = True
  print ("Received: " + bytes.decode(e.text))

def fireConnectionConnected(e):
  pass

print("********************************************************************")
print("* This is a demo to show how to connect to an Azure Relay Service  *")
print("* in order to send and receive data.                               *")
print("********************************************************************")

try:
  azurerelaysender1 = AzureRelaySender()
  azurerelaysender1.on_data_in = fireDataIn
  azurerelaysender1.on_ssl_server_authentication = fireSSLServerAuthentication
  azurerelaysender1.on_connection_status = fireConnectionStatus
  azurerelaysender1.on_connection_connected = fireConnectionConnected
  
  azurerelaysender1.set_access_key(input("Set Access Key: "))
  azurerelaysender1.set_access_key_name(input("Set Access Key Name: "))
  azurerelaysender1.set_namespace_address(input("Set Namespace Address: "))
  azurerelaysender1.set_hybrid_connection("hc1")
  print("\n")
  
  azurerelaysender1.connect()
  print("Connected")
  while True: 
    command = int(input("Please input command: 1 [Send Data] or  2 [Exit]: "))
    if command == 1:
      send = input("Please enter data to send: ")
      azurerelaysender1.send(send.encode())
      dataReceived = False
      print("Now waiting for response...")
      while dataReceived == False:
        azurerelaysender1.do_events()
    else:
      azurerelaysender1.disconnect()
      break
      
except IPWorksMQError as e:
  print("ERROR: %s" % e.message)

